### PR TITLE
Added a setMatrix() function to set the matrix of the mesh shader. 

### PR DESCRIPTION
**Pull request by ggarra13**
This keeps the draw..() routines as they were.

It can be extended to also modify other shaders too.

### CODE REVIEW

File: lib/tlTimeline/IRender.h [hunk 1]
```diff
@@ -48,6 +48,9 @@ namespace tl
             //! Finish a render.
             virtual void end() = 0;
 
+            //! Sets the mesh matrix to a new value
+            virtual void setMatrix(const math::Matrix4x4f&) = 0;
```
Comment (darbyjohnston) [inline lib/tlTimeline/IRender.h:52]: I don't think having "setMatrix()" is enough, there is no way of returning it to the previous state. Maybe what we need is "pushMatrix()" and "popMatrix()" so the matrix can be modified and then reverted. The matrix should also apply to all of the primitives for consistency, or there should at least be a note in the comments about the limitations.

Comment (ggarra13) [reply lib/tlTimeline/IRender.h:52]: I was unsure what shaders are used only in the display pipeline and what shaders might be used both in the FBO texture creation as well as in the viewport rendering/baking.  I thought only the "mesh" shader was used.

I think that the pushMatrix/popMatrix functions is overkill.  You'd be emulating the opengl2 functionality.

Comment (darbyjohnston) [reply lib/tlTimeline/IRender.h:52]: Probably all of the drawing functions could be rewritten using the mesh functionality, that would reduce the complexity and provide a common code path.

Adding "pushMatrix()" and "popMatrix()" may be overkill, but I don't think the "setMatrix()" function is sufficient for the reasons I gave above.

Comment (ggarra13) [reply lib/tlTimeline/IRender.h:52]: Okay.  I'll add a pushMatrix() and popMatrix() functionality.  That means I need to create Shader::getUniform() functions first, right?

**Comment by darbyjohnston** [comment]
Could you attach a sketch or screenshot of what you would like to draw? I'm still a bit unclear about it and maybe there is another way to achieve the result. I'm trying to be very careful with the render API since one of my next tasks will be porting it to other graphics APIs, and to minimize the work I would like to keep it as minimal as possible.

**Comment by ggarra13** [comment]
> El 17 nov. 2022, a las 14:41, Darby Johnston ***@***.***> escribió:
> 
> 
> Could you attach a sketch or screenshot of what you would like to draw? I'm still a bit unclear about it and maybe there is another way to achieve the result. I'm trying to be very careful with the render API since one of my next tasks will be porting it to other graphics APIs, and to minimize the work I would like to keep it as minimal as possible.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***>

Find attached a picture of the drawings I am doing with the modified api.  Note that the drawing of shapes extends beyond the canvas and follows the panning and zooming around.  Also, the color green does not get picked by the pixel toolbar when the cursor is on one of the drawings.


—
Gonzalo Garramuno
***@***.***

**Comment by darbyjohnston** [comment]
Thanks for the screenshot, do you also have one showing the use of the rectangle outlines?

I think I understand the issue now but would like to think more about a general solution. I need to finish up some other work first, so maybe we can continue this discussion next week.

**Comment by ggarra13** [comment]
Find two screenshots attached.  The image is 4K and it is resized by the panel to the right.
The image with the rectangle (two lines only shown) is the one drawn with a mesh.  The one with the outline perfectly drawn is the one with GL_LINE_LOOP.
Captura de pantalla 2022-11-17 a la(s) 18 21 54
Captura de pantalla 2022-11-17 a la(s) 18 23 37

**Comment by darbyjohnston** [comment]
Hi, and sorry for the delay! I've been working on some improvements to the renderer in the "scene" branch and I think I'm pretty close to supporting this functionality. I've added a `setTransform()` function that takes a 4x4 matrix that applies to all of the primitives including the image/video rendering. I've also added functions for adjusting the viewport and setting a clip rectangle which are useful when rendering UIs. I'm currently testing it out, and maybe later this week if you have time you can try it out for your use case. Thanks!

**Pull request closed without merging**